RFP Fields (matching, bounty, dates)

wbobeirne opened the pull request:
Closes #84.

### What This Does

* Adds `matching`, `bounty`, `date_closes`, `date_opened`, and `date_closed` to the RFP model.
* Adds admin fields for setting matching, bounty, and date_closes.
* Displays the above 3 on RFPs
* Proposals submitted against RFPs with matching automatically set `contribution_matching = 1.0`
* Proposals made from RFPs no longer take the title and brief
* RFPs will have their status soft-set to `CLOSED` after `date_closes`. This won't be reflected in the db, but will show up everywhere else. Potentially shift this to a task so that it's updated in the db.

### Steps to Test

* Update your db, confirm that your old RFPs were given `date_opened` and `date_closed` fields correctly.
* Create a new RFP, confirm that the new fields save and display properly.
* Create a new proposal off of the RFP, confirm that only the category is inherited and the `contribution_matching` was correctly set.
* Create an RFP with the `date_closes` set in the past, confirm it shows as `CLOSED`.

## Screenshots

### List display with closes date

<img width="821" alt="screen shot 2019-02-08 at 1 53 06 pm" src="https://user-images.githubusercontent.com/649992/52500711-a7237980-2bac-11e9-8b56-8b77bc02f3e0.png">

<br/>
<br/>

#### Detail display showing new fields at the bottom

<img width="874" alt="screen shot 2019-02-08 at 2 04 12 pm" src="https://user-images.githubusercontent.com/649992/52500721-ae4a8780-2bac-11e9-8a6d-40fde017bdbe.png">

<br/>
<br/>

#### New admin fields

<img width="919" alt="screen shot 2019-02-08 at 2 17 51 pm" src="https://user-images.githubusercontent.com/649992/52500732-b276a500-2bac-11e9-9813-fee75e34a28b.png">

<br/>
<br/>

#### Status field when date_closes is in the past

<img width="423" alt="screen shot 2019-02-08 at 2 18 13 pm" src="https://user-images.githubusercontent.com/649992/52500755-bdc9d080-2bac-11e9-89b9-734418cc02b4.png">

dternyak commented:
I'm having some trouble upgrading my database. Tried with the last database state and after dropping all my tables, same result.

```
...
  File "/Users/danielternyak2/ActiveDevelopment/zcash-grant-system/backend/venv/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) table "rfp_proposal" does not exist
 [SQL: '\nDROP TABLE rfp_proposal'] (Background on this error at: http://sqlalche.me/e/f405)
```

dternyak commented:
I went ahead and blew away the migrations locally and re-migrated from scratch to keep the ball rolling.

Have a few high-level comments:

- I don't see any indicator on the proposal detail view that a bounty is being provided against the proposal when the admin has set one
- The toggle to match doesn't appear to activate until I edit the approved proposal *after* it has been published. Toggling it at the RFP view doesn't seem to have an affect (maybe I'm doing something wrong?)
- Very nitpicky: A better place to show that a proposal is being made against a request might be near the title section? Here's my attached PRO example of where it might go. 

<img width="1130" alt="screen shot 2019-02-09 at 4 38 30 pm" src="https://user-images.githubusercontent.com/7861465/52527148-53d32900-2c89-11e9-8ed7-8c3992680845.png">

AMStrix commented:
I had a fresh DB from `develop` with one RFP without a proposal, and a single created proposal using `flask create-proposals 1` and got this `flask db upgrade` error:

```
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) table "rfp_proposal" does not exist
 [SQL: '\nDROP TABLE rfp_proposal'] (Background on this error at: http://sqlalche.me/e/f405)
Sentry is attempting to send 1 pending error messages
Waiting up to 2.0 seconds
Press Ctrl-C to quit
```

AMStrix commented:
Upon clearing out my database completely and running `flask db upgrade` I get:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> 4af29f8b2143, empty message
INFO  [alembic.runtime.migration] Running upgrade 4af29f8b2143 -> e0d970ed6500, empty message
INFO  [alembic.runtime.migration] Running upgrade e0d970ed6500 -> 722b4e7f7a58, empty message
INFO  [alembic.runtime.migration] Running upgrade 722b4e7f7a58 -> eddbe541cff1, empty message
INFO  [alembic.runtime.migration] Running upgrade eddbe541cff1 -> edf057ef742a, empty message
INFO  [alembic.runtime.migration] Running upgrade edf057ef742a -> fa1fedf4ca08, empty message
INFO  [alembic.runtime.migration] Running upgrade fa1fedf4ca08 -> 310dca400b81, empty message
INFO  [alembic.runtime.migration] Running upgrade 310dca400b81 -> d39bb526eef4, Adds RFP bounty, matching, and date fields
Traceback (most recent call last):
  File "/Users/aaron/Documents/git/zcash-grant-system/backend/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1193, in _execute_context
    context)
  File "/Users/aaron/Documents/git/zcash-grant-system/backend/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 509, in do_execute
    cursor.execute(statement, parameters)
psycopg2.ProgrammingError: table "rfp_proposal" does not exist

...

sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) table "rfp_proposal" does not exist
 [SQL: '\nDROP TABLE rfp_proposal'] (Background on this error at: http://sqlalche.me/e/f405)
Sentry is attempting to send 1 pending error messages
```

I'm do a fresh migration so I can test.

AMStrix commented on backend/grant/rfp/models.py:
```diff
@@ -82,15 +113,39 @@ class Meta:
+    def get_date_opened(self, obj):
+        return dt_to_unix(obj.date_opened) if obj.date_opened else None
+    
+    def get_date_closed(self, obj):
```
OT, but I while its in my sieve-like mind, I wonder if [custom fields](https://marshmallow.readthedocs.io/en/3.0/custom_fields.html) would be a good match for these.

wbobeirne commented:
Good calls on the bunk migration and toggle problems y'all. Not sure about the former, but I think I know what's up with the latter. I'll get them fixed shortly.

@dternyak as to your comments:

> I don't see any indicator on the proposal detail view that a bounty is being provided against the proposal when the admin has set one

I didn't think this needed any visual treatment since the ZF should just send a contribution as soon as it gets approved, right? Not really sure we ever ironed out how bounties work, exactly.

> Very nitpicky: A better place to show that a proposal is being made against a request might be near the title section? Here's my attached PRO example of where it might go.

I thought about that spot but it doesn't work well for long titles, or as you shrink to go responsive. I also thought about putting it under the title, but then the alignment with the sidebar goes out of whack. I'm definitely down to rework where this goes in another issue though, I'm far from happy with it either.

dternyak commented:
> I didn't think this needed any visual treatment since the ZF should just
send a contribution as soon as it gets approved, right? Not really sure we
ever ironed out how bounties work, exactly.


Definitely, these were never exactly specified.

I think it would be valuable to show to the community that the foundation
is directly funding a portion of the fund, and then still allow users to
contribute. we shouldn't need the progress bar anymore though, since the
bounty is gauranteed upon delivery.
On Thu, Feb 14, 2019 at 1:51 PM wbobeirne <notifications@github.com> wrote:

> Good calls on the bunk migration and toggle problems y'all. Not sure about
> the former, but I think I know what's up with the latter. I'll get them
> fixed shortly.
>
> @dternyak <https://github.com/dternyak> as to your comments:
>
> I don't see any indicator on the proposal detail view that a bounty is
> being provided against the proposal when the admin has set one
>
> I didn't think this needed any visual treatment since the ZF should just
> send a contribution as soon as it gets approved, right? Not really sure we
> ever ironed out how bounties work, exactly.
>
> Very nitpicky: A better place to show that a proposal is being made
> against a request might be near the title section? Here's my attached PRO
> example of where it might go.
>
> I thought about that spot but it doesn't work well for long titles, or as
> you shrink to go responsive. I also thought about putting it under the
> title, but then the alignment with the sidebar goes out of whack. I'm
> definitely down to rework where this goes in another issue though, I'm far
> from happy with it either.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/grant-project/zcash-grant-system/pull/179#issuecomment-463768079>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf02SvMF_F20Sm7sQl8iuWQ93LRUFcUks5vNb45gaJpZM4axMeO>
> .
>

dternyak commented:
> I didn't think this needed any visual treatment since the ZF should just
send a contribution as soon as it gets approved, right? Not really sure we
ever ironed out how bounties work, exactly.


Definitely, bounties were never exactly specified.

The way I had envisioned it is with a slightly modified campaign block to show to the community that the foundations directly funding a portion of the fund, but still allow users to
contribute/crowdfund. 

We shouldn't need the progress bar anymore though, since the
bounty is guaranteed upon delivery. We may want to still separately show much the foundation is promising + how much the community has independently contributed however.

wbobeirne commented:
Okee doke, merged in develop, fixed the `matching: false` bug, and fixed the migration (Looks like 2 migrations were dropping the same table, womp womp.) Should be good to go.

dternyak commented:
👌 Going to give this another run through and make sure it's all good, then merging.

AMStrix commented:
Smooooooth.
